README.md
automaid -- a Python package to process MERMAID files

Originally written by Sebastien Bonnieux.
Later maintained by Joel D. Simon and Frederik J. Simons.
Currently maintained and extended by Joel D. Simon (jdsimon@princeton.edu) for EarthScope-Oceans.

This program converts raw data transmitted by Mermaid instruments to classify
their data, correct clock drifts, interpolate float positions, generate seismic
SAC and miniSEED files, plot seismic events and dives, and generate KML files.

### 1. INSTALLATION

This installation procedure has been tested with MacOS. For Linux the
procedure is valid but one could prefer to use a package manager.
For Windows the installation of Python 2.7 is valid but the
compilation of the wavelet inversion program with "make" could be
problematic.

An easy installation procedure is described here:

* Install [Miniconda](https://conda.io/miniconda.html) or
  [Anaconda](https://www.anaconda.com/download/) (which requires more
  disk space). (You may already have it, you might have to do `module
  load anaconda/5.2.0` to specify the precise version).
* Restart your terminal to load the new PATH variables.
* Add the conda-forge channel:
  `conda config --add channels conda-forge`
* Create a virtual environment called "pymaid":
  `conda create -n pymaid python=2.7`

* Make sure you are in the `bash` shell!

* Activate the environment:
  `source activate pymaid`
* Install obspy:
  `conda install obspy`
* Install plotly 2.7.0:
  `conda install plotly=2.7.0`
* Quit the virtual environment:
  `source deactivate`

In addition to the Python 2.7 installation it is necessary to compile,
using `make` the wavelet inversion programs located in
`scripts/src/V103/` and `scripts/src/V103EC/`. The compiled binaries
must be in the "bin" directory and must be named `icdf24_v103_test` and
`icdf24_v103ec_test`.

Finally, ensure the environmental variables:
(1) $MERMAID, itself a directory containing "server/" and "processed/", and
(2) $AUTOMAID, the path to this cloned directory, containing "scripts/",
are defined in your system.

### 2. USAGE

To use the application:

* Copy files from your Mermaid server into the "$MERMAID/server" directory:
  `scp username@host:\{"*.LOG","*.MER","*.vit"\} server`
* Activate the virtual environment:
  `source activate pymaid` or `conda activate pymaid` or (if "conda" not found)
  e.g., `source /Users/joelsimon/anaconda3/etc/profile.d/conda.sh ; conda  activate pymaid`
* Run the main.py file:
  `python $AUTOMAID/scripts/main.py`
* Quit the virtual environment:
  `source deactivate`

You will be getting the processed files into the directory `$MERMAID/processed`.
You may have to remove some error-prone log files and create some directories -
we will be editing the script for increased versatility as we go along.

The "main.py" file can be edited to select some options:

* A date range between which to process the data can be chosen with
the `begin` and `end` variables.
* A "redo" flag can be set to True in order to restart the processing
of data for each launch of the script. This flag force the deletion
of the content of the content of the `processed` directory.
* A `events_plotly` flag allow the user to plot interactive figures
of events in a html page. This kind of plot can be disabled to save
disk space.

An additional tool is available to invert a single ".MER" files. For
this, go in the `scripts` directory. Put a Mermaid file with the
extension ".MER" in the `scripts/tool_invert_mer` directory. And
finally run the script: `python tool_invert_mer.py`.


changelog.txt
Current:
*v3.3.0: 08-Dec-2020

    "Modify interpolation algorithm to return input location AND date when
     interpolation not performed, and write float32 metadata files"

-- Modify gps.linear_interpolation to return the input date and location, as
   opposed to the requested date and input location, when only a single GPS
   instance is retained (i.e., interpolation can not proceed); see esp. commit
   1ee1ca269bf27059f860c1e1385b495cea81526c
-- Write gps.csv alongside previous gps.txt
-- Collect ALL and ONLY mSEED/SAC metadata defined by automaid in
   event.obspy_trace_stats attribute dictionary, with the understanding that
   ObsPy DOES NOT automatically cast them into float32; that must be done when
   writing to csv/txt
-- Define new metadata fields: elev(STEL), azimuth(CMPAZ), SACdip(CMPINC)
-- Write mseed2sac_metadata.csv/txt
-- Write automaid_metadata.csv/txt
-- Remove dives.attach_is_complete_mer_to_dive_events, whose utility was
   superseded by dive.is_complete_mer_file attribute in v1.4

Previous:
*v3.2.0: 11-Nov-2020
  -- Parse/zero-pad Osean-defined serial number into five-character station name
     (KSTNM) and "generic name of recording instrument" (KINST) in dives; add
     those fields to the SAC header (and default others) in events
  -- Give finer control of start/end dates of analysis ("filterDate") by
     specifically checking individual GPS fixes within LOG/MER files as opposed
     to accepting or rejecting entire files (that contain multiple points) based
     only on their file names; update start dates of Princeton-owned floats to
     first date in the LOG after GPS synchronization (internal clock zeroed)
  -- Write loc.txt file of interpolated STLA and STLO (station locations); same
     data that go in STLA and STLO SAC-header fields
  -- Add dictionary of raw strings (not processed) that collect GPS parameters
     from LOG and MER files in their (odd and different) native formats (e.g.,
     degrees + decimal minutes) to each GPS instance and write those strings to
     gps.txt
  -- Remove Osean-serial number (e.g., '452.020-P-25') from text files that I
     write (e.g., gps.txt) to differentiate them from files written by MERMAID;
     add automaid version number and source-code url (ESO github) header to each

*v3.1.0: 05-Nov-2020
  -- Allow single GPS for surface-drift interpolation (previously,
     gps.linear_interpolation failed if len < 2)
  -- Validate .MER file reported in .LOG file exists before processing
     (sometimes the .MER file hits the server much later than the associated
     .LOG)
  -- Add previous dive information to printout and output *traces.txt

*v3.0.0: 03-Nov-2020
   -- Make GPS list extension more strict: only appends previous(next) dive's
      GPS list if that .LOG represents a complete dive AND the appropriate
      after(before)-dive GPS list exists, OR the previous(next) .LOG does not
      represent a dive (like an error or a reboot) but it still contains legit
      GPS points that we can use (see especially commit
      14f871e12194230e3d897bc5577ee65168c0f770 message); this avoids the
      potential pitfall of v2.X.X where, for example, the GPS list before the
      dive of the previous dive is attached, meaning we are attempting to
      interpolate from two dives back in time
   -- Removed pickling from main.py

*v2.1.0: 02-Nov-2020
  -- Attach interpolation dictionary to every GPS instance so that interpolation
     parameters may be teased apart; write output *gps_interpolation.txt file to
     explain the drifts computed in the three regimes (descent, drift, ascent)
  -- Swap default class attributes for default instance attributes to avoid
     unintended global overwrite of non-assigned class attributes:
	 class SomeClass:
	     cls_attr = foo
		  def __init__():
		      ...
	 class SomeClass:
	     def __init__():
		 self.inst_attr = foo
  -- Allow abbreviated '-s' and '-p' optional input flags
  -- Rewrite interpolations in Dives and GPS in plain English -- this does not
     change the algorithm at all (but it makes it much more verbose but
     hopefully more readable and easier to debug?)
  -- Add writing of last_dive_pressure_offset.txt to alert when (corrected)
     external pressure is near/beyond the 300 mbar limit
  -- Add writing of XXX.XXX-P-??_dives.txt, which contains the same info as the
     printout that ends each float loop
  -- Attach .mer_environment_name and .mer_binary_name to each Event instance
     (which is also contained in the encapsulating Dive instance, but this make
     access more explicit), and change .mmd_data_name to .mer_binary_name to
     distinguish it from the .MER file associated with the environment (which
     may be different)
  -- Modify attribute names (e.g. "self.mmd..." for "self.mer..") for , which
     should be fine at this point because automaid is still self contained, as
     far as I'm aware

*v2.0.0: 16-Oct-2020
  -- Utilize complete GPS list from both .LOG and .MER files to compute
     interpolated station locations (previously, the .MER was searched first, and
     if no GPS fixes existed there, the .LOG was used)
  -- Explicitly separate newly-combined .LOG/.MER GPS list into "before_dive",
     "after dive" etc. sublists rather than relying on list-slicing that may or
     may not be valid depending transmission state (Iridium/GPS failures and
     drops can result in incomplete files, meaning that the final GPS point,
     which should be after the dive, may not have been transmitted,
     potentially(?) resulting in the actual final GPS point in the file being
     before the dive)
  -- Add header lines to output GPS text files
  -- Generates 45 additional SAC/mSEED files for 16 Princeton floats due to
     increased number of GPS fixes now attached to each Dive instance and
     available for GPS interpolation

*v1.4.0: 14-Oct-2020
  -- Check the actual byte length of each event contained in every .MER file,

scripts/main.py
# automaid -- a Python package to process MERMAID files
# pymaid environment (Python v2.7)
#
# Usage: python main.py [-p processed] [-s server]
#
# Original author: Sebastien Bonnieux
# Current maintainer: Joel D. Simon (JDS)
# Contact: jdsimon@alumni.princeton.edu | joeldsimon@gmail.com
# Last modified by JDS: 08-Dec-2020, Python 2.7.15, Darwin-18.7.0-x86_64-i386-64bit

import os
import argparse
import shutil
import glob
import datetime
import dives
import events
import vitals
import kml
import gps
import setup
import re
import utils

# Get current version number.
version = setup.get_version()

# Set depth of mixed layer (in meters) for drift interpolation
mixed_layer_depth_m = 50

# Set default paths
automaid_path = os.environ["AUTOMAID"]
def_mermaid_path = os.environ["MERMAID"]
def_server_path = os.path.join(def_mermaid_path, "server")
def_processed_path = os.path.join(def_mermaid_path, "processed")

# Parse (optional) command line inputs to override default paths
parser = argparse.ArgumentParser()
# problem: metavar=''   prints: "-s , --server"
# problem: metavar='\b' prints: "-s, --server", but misaligns the help statement...
parser.add_argument('-s',
                    '--server',
                    default=def_server_path,
                    dest='server',
                    #metavar='\b',
                    help="server directory (default: {:s})".format(def_server_path))
parser.add_argument('-p',
                    '--processed',
                    default=def_processed_path,
                    dest='processed',
                    #metavar='',
                    help="processed directory (default: {:s})".format(def_processed_path))
args = parser.parse_args()
server_path = os.path.abspath(args.server)
processed_path = os.path.abspath(args.processed)

# Set an inclusive time range of analysis for a specific float
filterDate = {
    "452.112-N-01": (datetime.datetime(2018, 12, 27), datetime.datetime(2100, 1, 1)),
    "452.112-N-02": (datetime.datetime(2018, 12, 28), datetime.datetime(2100, 1, 1)),
    "452.112-N-03": (datetime.datetime(2018, 4, 9), datetime.datetime(2100, 1, 1)),
    "452.112-N-04": (datetime.datetime(2019, 1, 3), datetime.datetime(2100, 1, 1)),
    "452.112-N-05": (datetime.datetime(2019, 1, 3), datetime.datetime(2100, 1, 1)),
    "452.020-P-06": (datetime.datetime(2018, 6, 26), datetime.datetime(2100, 1, 1)),
    "452.020-P-07": (datetime.datetime(2018, 6, 27), datetime.datetime(2100, 1, 1)),
    # *
    "452.020-P-08": (datetime.datetime(2018, 8,  5, 13, 23, 14), datetime.datetime(2100, 1, 1)),
    "452.020-P-09": (datetime.datetime(2018, 8,  6, 15, 21, 26), datetime.datetime(2100, 1, 1)),
    "452.020-P-10": (datetime.datetime(2018, 8,  7, 12, 53, 42), datetime.datetime(2100, 1, 1)),
    "452.020-P-11": (datetime.datetime(2018, 8,  9, 11,  2,  6), datetime.datetime(2100, 1, 1)),
    "452.020-P-12": (datetime.datetime(2018, 8, 10, 19, 51, 31), datetime.datetime(2100, 1, 1)),
    "452.020-P-13": (datetime.datetime(2018, 8, 31, 16, 50, 23), datetime.datetime(2100, 1, 1)),
    "452.020-P-16": (datetime.datetime(2018, 9,  4,  7, 12, 15), datetime.datetime(2100, 1, 1)),
    "452.020-P-17": (datetime.datetime(2018, 9,  4, 11,  2, 54), datetime.datetime(2100, 1, 1)),
    "452.020-P-18": (datetime.datetime(2018, 9,  5, 17, 38, 32), datetime.datetime(2100, 1, 1)),
    "452.020-P-19": (datetime.datetime(2018, 9,  6, 20,  7, 30), datetime.datetime(2100, 1, 1)),
    "452.020-P-20": (datetime.datetime(2018, 9,  8, 10, 32,  8), datetime.datetime(2100, 1, 1)),
    "452.020-P-21": (datetime.datetime(2018, 9,  9, 17, 42, 36), datetime.datetime(2100, 1, 1)),
    "452.020-P-22": (datetime.datetime(2018, 9, 10, 19,  7, 21), datetime.datetime(2100, 1, 1)),
    "452.020-P-23": (datetime.datetime(2018, 9, 12,  2,  4, 14), datetime.datetime(2100, 1, 1)),
    "452.020-P-24": (datetime.datetime(2018, 9, 13,  8, 52, 18), datetime.datetime(2100, 1, 1)),
    "452.020-P-25": (datetime.datetime(2018, 9, 14, 11, 57, 12), datetime.datetime(2100, 1, 1)),
    # *
    "452.020-P-0050": (datetime.datetime(2019, 8, 11), datetime.datetime(2100, 1, 1)),
    "452.020-P-0051": (datetime.datetime(2019, 7, 1), datetime.datetime(2100, 1, 1)),
    "452.020-P-0052": (datetime.datetime(2019, 7, 1), datetime.datetime(2100, 1, 1)),
    "452.020-P-0053": (datetime.datetime(2019, 7, 1), datetime.datetime(2100, 1, 1)),
    "452.020-P-0054": (datetime.datetime(2019, 7, 1), datetime.datetime(2100, 1, 1))
}
# *I found dates in the same range (~minutes before) as misalo.txt and set these filterDates to the
# actual corresponding date in the LOG; if the date did not match exactly I looked for the first
# date where the clock drift reset and the associated LOG recorded an actual dive

# Boolean set to true in order to delete every processed data and redo everything
redo = False

# Figures commented by default (take heaps of memory)
events_plotly = False
events_mseed = True
events_sac = True
events_png = False

# Dictionary to save data in a file
dives_dict = dict()

def main():
    # Set working directory in "scripts"
    os.chdir(os.path.join(automaid_path, "scripts", ""))

    # Create processed directory if it doesn't exist
    if not os.path.exists(processed_path):
        os.mkdir(processed_path)

    # Search Mermaid floats
    vitfile_path = os.path.join(server_path, "*.vit")
    mfloats = [p.split("/")[-1][:-4] for p in glob.glob(vitfile_path)]

    # Initialize empty dict to hold the instance of every last complete dive for
    # every MERMAID
    lastdive = dict()

    # For each Mermaid float
    for mfloat in mfloats:
        mfloat_serial = mfloat[-4:]
        print("Processing {:s} .LOG & .MER files...".format(mfloat_serial))

        # Set the path for the float
        mfloat_path = os.path.join(processed_path, mfloat, "")

        # Get float number
        mfloat_nb = re.findall("(\d+)$", mfloat)[0]

        # Delete the directory if the redo flag is true
        if redo and os.path.exists(mfloat_path):
            shutil.rmtree(mfloat_path)

        # Create directory for the float
        if not os.path.exists(mfloat_path):
            os.mkdir(mfloat_path)

        # Remove existing files in the processed directory (the script may have been previously
        # executed, copied the files, then failed)
        for f in glob.glob(mfloat_path + "*.*"):
            os.remove(f)

        # Copy appropriate files in the directory and remove files outside of the time range
        files_to_copy = list()
        extensions = ["000", "001", "002", "003", "004", "005", "LOG", "MER"]
        for extension in extensions:
            files_to_copy += glob.glob(os.path.join(server_path, mfloat_nb +  "*." + extension))

        # Add .cmd, .out, and .vit files
        files_to_copy += glob.glob(os.path.join(server_path, mfloat + "*"))

        # Copy files
        for f in files_to_copy:
            shutil.copy(f, mfloat_path)

        # Really: collect all the .MER files (next we correlate their environments to .LOG files)
        print(" ...compiling a list of events from {:s} .MER files (GPS & seismic data)..." \
              .format(mfloat_serial))
        mevents = events.Events(mfloat_path)

        # Determine the time range of analysis (generally; birth to death of a MERMAID)
        if mfloat in filterDate.keys():
            begin = filterDate[mfloat][0]
            end = filterDate[mfloat][1]
        else:
            begin = datetime.datetime(1000, 1, 1)
            end = datetime.datetime(3000, 1, 1)

        # Really: collect all the .LOG files in order (1 .LOG == 1 Dive)
        print(" ...matching those events to {:s} .LOG ('dive') files (GPS & dive metadata)..." \
              .format(mfloat_serial))
        mdives = dives.get_dives(mfloat_path, mevents, begin, end)

        # Generate logs and plots for each dive
        for dive in mdives:
            # Create the directory
            if not os.path.exists(dive.export_path):
                os.mkdir(dive.export_path)
            # Generate log
            dive.generate_datetime_log()
            # Generate mermaid environment file
            dive.generate_mermaid_environment_file()
            # Generate dive plot
            dive.generate_dive_plotly()

        # Compute clock drift correction for each event
        for dive in mdives:
            dive.correct_events_clockdrift()

        # Interpolate for the locations that MERMAID passed out of/in to the surface and mixed
        # layers, and where it was when it recorded any events associated with the dive
        mdives[0].compute_station_locations(None, mdives[1], mixed_layer_depth_m)
        i = 1
        while i < len(mdives)-1:
            mdives[i].compute_station_locations(mdives[i-1], mdives[i+1], mixed_layer_depth_m)
            i += 1

        # We skipped event computation for the last dive (awaiting the next dive's GPS) which also
        # adds a reference to the previous dive; correct that here for the last Dive instance
        if len(mdives) > 1:
            mdives[-1].prev_dive_log_name = mdives[-2].log_name
            mdives[-1].prev_dive_mer_environment_name = mdives[-2].mer_environment_name

        # Generate plots, SAC, and miniSEED files for each event
        print(" ...writing {:s} sac/mseed/png/html output files...".format(mfloat_serial))
        for dive in mdives:
            if events_png:
                dive.generate_events_png()
            if events_plotly:
                dive.generate_events_plotly()
            if events_sac:
                dive.generate_events_sac()
            if events_mseed:
                dive.generate_events_mseed()

        # Plot vital data
        kml.generate(mfloat_path, mfloat, mdives)
        vitals.plot_battery_voltage(mfloat_path, mfloat + ".vit", begin, end)
        vitals.plot_internal_pressure(mfloat_path, mfloat + ".vit", begin, end)
        vitals.plot_pressure_offset(mfloat_path, mfloat + ".vit", begin, end)
        if len(mdives) > 1:
            vitals.plot_corrected_pressure_offset(mfloat_path, mdives, begin, end)

        # Write csv and txt files containing all GPS fixes from .LOG and .MER
        gps.write_gps(mdives, processed_path, mfloat_path)

        # Write text file detailing event-station location interpolation parameters
        gps.write_gps_interpolation_txt(mdives, processed_path, mfloat_path)

        # Write helpful printout detailing every dive, and how .LOG and .MER
        # files connect
        dives.generate_printout(mdives, mfloat_serial)

        # Write text file detailing how .LOG and .MER files connect
        dives.write_dives_txt(mdives, processed_path, mfloat_path)

        # Write a text file relating all SAC and mSEED to their associated .LOG
        # and .MER files
        events.write_traces_txt(mdives, processed_path, mfloat_path)

        # Write a text file with our best-guess at the location of MERMAID at the time of recording
        events.write_loc_txt(mdives, processed_path, mfloat_path)

        # Write mseed2sac and automaid metadata csv and text files
        events.write_metadata(mdives, processed_path, mfloat_path)

        # Clean directories
        for f in glob.glob(mfloat_path + "/" + mfloat_nb + "_*.LOG"):
            os.remove(f)
        for f in glob.glob(mfloat_path + "/" + mfloat_nb + "_*.MER"):
            os.remove(f)

        # Add dives to growing dict
        dives_dict[mfloat] = mdives

    # Done looping through all dives for each float
    #______________________________________________________________________________________#

    # Print a text file of corrected external pressures measured on the final
    # dive, and warn if any are approaching the limit of 300 mbar (at which
    # point adjustment is required)
    vitals.write_corrected_pressure_offset(dives_dict, processed_path)

if __name__ == "__main__":
    main()
